File: src/getml-app/src/tcp/progressLogger.go
// Copyright 2024 Code17 GmbH
// 
// This file is licensed under the Elastic License 2.0 (ELv2). 
// Refer to the LICENSE.txt file in the root of the repository 
// for details.
// 

package tcp

import (
	"errors"
	"strconv"
	"strings"
)

type progressLogger struct {
	logger *logger

	numIterations int

	totalNumIterations int
}

func newProgressLogger(
	logger logger,
	totalNumIterations int) (*progressLogger, error) {

	if totalNumIterations <= 0 {
		return nil, errors.New(
			"'totalNumIterations' needs to be greater than 0")
	}

	return &progressLogger{
		logger:             &logger,
		numIterations:      0,
		totalNumIterations: totalNumIterations,
	}, nil

}

func (l *progressLogger) log(msg string) {
	if strings.Contains(msg, "Progress:") {
		l.increment()
	}
}

func (l *progressLogger) increment() {
	l.numIterations++
	progress := (100 * l.numIterations) / l.totalNumIterations
	l.logger.log("Progress: " + strconv.Itoa(progress) + "%")
}

func (l *progressLogger) numRemaining() int {

	if l.numIterations > l.totalNumIterations {
		return 0
	}

	return l.totalNumIterations - l.numIterations

}

File: src/engine/src/logging/ProgressLogger.cpp
// Copyright 2024 Code17 GmbH
//
// This file is licensed under the Elastic License 2.0 (ELv2).
// Refer to the LICENSE.txt file in the root of the repository
// for details.
//

#include "logging/ProgressLogger.hpp"

namespace logging {

ProgressLogger::ProgressLogger(
    const std::string& _msg,
    const std::shared_ptr<const AbstractLogger>& _logger, const size_t _total,
    const size_t _begin, const size_t _end)
    : begin_(_begin),
      current_value_(0),
      end_(_end),
      logger_(_logger),
      total_(_total) {
  if (logger_ && total_ > 0 && _msg != "") {
    logger_->log(_msg);
  }
}

}  // namespace logging
